### ISSUE TITLE
Uncaught exception in Firebase Database runloop

### ISSUE THREAD

**Issue by zoltish** [firebase/firebase-android-sdk#3276]
### [REQUIRED] Step 2: Describe your environment

  * Android Studio version: Android Studio Bumblebee | 2021.1.1 Beta 5 Build #AI-211.7628.21.2111.7956428
  * Firebase Component: Database
  * Component version: 20.0.2
    
    _implementation platform('com.google.firebase:firebase-bom:29.0.0')
    implementation 'com.google.firebase:firebase-auth-ktx'
    implementation 'com.google.firebase:firebase-database-ktx'_

#### Steps to reproduce:

When the user signs in, data from firebase is downloaded, and then any (other) local data is uploaded. It seems that if there is a lot of local data, the application eventually crashes like this. I can easily reproduce it, and its not an unreasonable amount of data (this occurs when using the sync in a real world scenario). 

I can share implementation details if you think that could help, just let me know what portion of it would be of interest. I cant share the entire project though, and as Im in the middle of a release I dont think Ill have time to create a repro-project anytime soon. If the issue is as straight-forward as it seems, just uploading enough data in one go should lead to this crash eventually. 

I am using `setPersistenceEnabled(true)` and Ill include a screenshot of what the data looks like below. The data is rather simple and contains a flattened structure, like /branch/userId/dataId/data; e.g. /sessions/myUserId/5/data. 

#### Relevant Code:

Stacktrace:
```
java.lang.RuntimeException: Uncaught exception in Firebase Database runloop (20.0.2). If you are not already on the latest version of the Firebase SDKs, try updating your dependencies. Should this problem persist, please file a report at https://github.com/firebase/firebase-android-sdk
        at com.google.firebase.database.android.AndroidPlatform$1$1.run(AndroidPlatform.java:96)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loopOnce(Looper.java:201)
        at android.os.Looper.loop(Looper.java:288)
        at android.app.ActivityThread.main(ActivityThread.java:7839)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1003)
     Caused by: java.lang.RuntimeException: java.lang.OutOfMemoryError: Failed to allocate a 24 byte allocation with 1023448 free bytes and 999KB until OOM, target footprint 201326592, growth limit 201326592; giving up on allocation because <1% of heap free after GC.
        at com.google.firebase.database.core.persistence.DefaultPersistenceManager.runInTransaction(DefaultPersistenceManager.java:244)
        at com.google.firebase.database.core.SyncTree.ackUserWrite(SyncTree.java:237)
        at com.google.firebase.database.core.Repo.ackWriteAndRerunTransactions(Repo.java:441)
        at com.google.firebase.database.core.Repo.access$800(Repo.java:61)
        at com.google.firebase.database.core.Repo$8.onRequestResult(Repo.java:479)
        at com.google.firebase.database.connection.PersistentConnectionImpl$5.onResponse(PersistentConnectionImpl.java:1281)
        at com.google.firebase.database.connection.PersistentConnectionImpl.onDataMessage(PersistentConnectionImpl.java:513)
        at com.google.firebase.database.connection.Connection.onDataMessage(Connection.java:167)
        at com.google.firebase.database.connection.Connection.onMessage(Connection.java:131)
        at com.google.firebase.database.connection.WebsocketConnection.appendFrame(WebsocketConnection.java:259)
        at com.google.firebase.database.connection.WebsocketConnection.handleIncomingFrame(WebsocketConnection.java:306)
        at com.google.firebase.database.connection.WebsocketConnection.access$500(WebsocketConnection.java:34)
        at com.google.firebase.database.connection.WebsocketConnection$WSClientTubesock$2.run(WebsocketConnection.java:86)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:920)
     Caused by: java.lang.OutOfMemoryError: Failed to allocate a 24 byte allocation with 1023448 free bytes and 999KB until OOM, target footprint 201326592, growth limit 201326592; giving up on allocation because <1% of heap free after GC.
        at com.google.firebase.database.collection.ArraySortedMap.insert(ArraySortedMap.java:103)
        at com.google.firebase.database.core.utilities.ImmutableTree.setTree(ImmutableTree.java:257)
        at com.google.firebase.database.core.CompoundWrite.addWrite(CompoundWrite.java:97)
        at com.google.firebase.database.core.WriteTree.layerTree(WriteTree.java:438)
        at com.google.firebase.database.core.WriteTree.resetTree(WriteTree.java:403)
        at com.google.firebase.database.core.WriteTree.removeWrite(WriteTree.java:158)
        at com.google.firebase.database.core.SyncTree$3.call(SyncTree.java:245)
        at com.google.firebase.database.core.SyncTree$3.call(SyncTree.java:238)
        at com.google.firebase.database.core.persistence.DefaultPersistenceManager.runInTransaction(DefaultPersistenceManager.java:239)
        at com.google.firebase.database.core.SyncTree.ackUserWrite(SyncTree.java:237) 
        at com.google.firebase.database.core.Repo.ackWriteAndRerunTransactions(Repo.java:441) 
        at com.google.firebase.database.core.Repo.access$800(Repo.java:61) 
        at com.google.firebase.database.core.Repo$8.onRequestResult(Repo.java:479) 
        at com.google.firebase.database.connection.PersistentConnectionImpl$5.onResponse(PersistentConnectionImpl.java:1281) 
        at com.google.firebase.database.connection.PersistentConnectionImpl.onDataMessage(PersistentConnectionImpl.java:513) 
        at com.google.firebase.database.connection.Connection.onDataMessage(Connection.java:167) 
        at com.google.firebase.database.connection.Connection.onMessage(Connection.java:131) 
        at com.google.firebase.database.connection.WebsocketConnection.appendFrame(WebsocketConnection.java:259) 
        at com.google.firebase.database.connection.WebsocketConnection.handleIncomingFrame(WebsocketConnection.java:306) 
        at com.google.firebase.database.connection.WebsocketConnection.access$500(WebsocketConnection.java:34) 
        at com.google.firebase.database.connection.WebsocketConnection$WSClientTubesock$2.run(WebsocketConnection.java:86) 
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:462) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) 
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:920) 
2021-12-29 06:58:22.986 ? E/Progression: (CrashlyticsUncaughtExceptionHandler.java:63) @DefaultDispatcher-worker-8 CrashlyticsUncaughtExceptionHandler#uncaughtException -> kotlin.Unit
    java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@6ffbcc0 rejected from com.google.firebase.database.core.utilities.DefaultRunLoop$1@768fd81[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 1806]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2086)
        at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:848)
        at java.util.concurrent.ScheduledThreadPoolExecutor.delayedExecute(ScheduledThreadPoolExecutor.java:334)
        at java.util.concurrent.ScheduledThreadPoolExecutor.schedule(ScheduledThreadPoolExecutor.java:562)
        at java.util.concurrent.ScheduledThreadPoolExecutor.execute(ScheduledThreadPoolExecutor.java:654)
        at com.google.firebase.database.core.utilities.DefaultRunLoop.scheduleNow(DefaultRunLoop.java:105)
        at com.google.firebase.database.core.Repo.scheduleNow(Repo.java:303)
        at com.google.firebase.database.DatabaseReference.setValueInternal(DatabaseReference.java:286)
        at com.google.firebase.database.DatabaseReference.setValue(DatabaseReference.java:159)
        at progression.firebase.api.FirebasePath.update-impl(FirebasePath.kt:94)
        at progression.firebase.api.FirebaseIntegration.upload(FirebaseIntegration.kt:222)
        at progression.firebase.api.FirebaseIntegration.access$upload(FirebaseIntegration.kt:23)
        at progression.firebase.api.FirebaseIntegration$upload$1.invokeSuspend(Unknown Source:15)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.internal.LimitedDispatcher.run(LimitedDispatcher.kt:39)
        at kotlinx.coroutines.scheduling.TaskImpl.run(Tasks.kt:95)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

Logs show how it falls apart, normally an operation takes ~300 ms; but that number grows exponentially rather quickly. I thought the sessions were the problem, but it fails when uploading ~350 exercises even. 
```
 -> upload Profiles (#1337): 368 ms
 -> upload Profiles (#1337): 347 ms
 -> upload Profiles (#1337): 363 ms
 -> upload Profiles (#1337): 340 ms
 -> upload Exercises (#9225629d-3fe3-4012-a2b7-95c1b04980a5): 504 ms
 -> upload Exercises (#-10): 867 ms
 -> upload Exercises (#9d816eaf-5159-4830-9ce4-64c0beee5c17): 898 ms
 -> upload Exercises (#d2043e07-9e38-499c-99d4-b79db70ba167): 930 ms
 -> upload Exercises (#-13): 947 ms
 -> upload Exercises (#8bef644b-b985-4335-aa7b-8a4631ed44c0): 978 ms
 -> upload Exercises (#-3): 1008 ms
 -> upload Exercises (#-11): 977 ms
 -> upload Exercises (#d22d4a32-9a96-43f2-81a7-3f6e9f079e9e): 1030 ms
 -> upload Exercises (#b3bce3b4-361d-4f7e-8ed1-ba3ae17f8c8a): 1394 ms
 -> upload Exercises (#792414db-1601-468a-b20e-1bdfb36b9d92): 1454 ms
 -> upload Exercises (#690e5687-a59d-47e1-b280-092efa9d2f0d): 1575 ms
 -> upload Exercises (#-6): 1626 ms
 -> upload Exercises (#0f744994-417b-4a6b-bcaa-6e50b8dfc5af): 1657 ms
 -> upload Exercises (#b052c7c1-339d-41ae-ade2-39d7ed65935f): 1772 ms
 -> upload Exercises (#-4): 1853 ms
 -> upload Exercises (#31e3a728-4582-4881-ad79-b581a4e7bba5): 1880 ms
 -> upload Exercises (#1d624ed4-b955-4277-85e8-1b5be595205b): 1990 ms
 -> upload Exercises (#-5): 2069 ms
 -> upload Exercises (#-2): 2172 ms
 -> upload Exercises (#e0b7ec9d-a77f-4e2b-b879-cf6ab426cd9a): 2196 ms
 -> upload Exercises (#c03b1e80-fe52-4b4c-b509-f27d14591b1e): 2289 ms
 -> upload Exercises (#91555941-8453-4340-92aa-be7eb9610160): 2315 ms
 -> upload Exercises (#c266a69f-171a-4853-896f-a37a3f65132e): 2393 ms
 -> upload Exercises (#1ee915ae-cfac-4b61-bc7b-203eed0ec686): 2454 ms
 -> upload Exercises (#9945f223-f5e5-466b-8e43-76762628a7eb): 2497 ms
 -> upload Exercises (#38f36e8a-aedb-4e85-b650-e406cd497431): 2531 ms
 -> upload Exercises (#ea69937e-f5ea-409e-a6f2-bace8ccc8869): 2598 ms
 -> upload Exercises (#-1): 2774 ms
 -> upload Exercises (#-9): 2849 ms
 -> upload Exercises (#12b3a778-db0f-4eb2-9b88-bd32e560c03d): 2952 ms
 -> upload Exercises (#17bd141e-e0ab-439d-89b5-5f3ce9342289): 3493 ms
 -> upload Exercises (#fea874a9-d0a9-4ea1-82f4-07257ca7adaf): 3690 ms
 -> upload Exercises (#68a0b252-c552-49b6-8657-dad5915b5ac4): 3758 ms
 -> upload Exercises (#d54d9934-341b-425b-9143-126c8282a65a): 3992 ms
 -> upload Exercises (#-8): 4092 ms
 -> upload Exercises (#8c91e488-e97b-44b6-b323-042924627c3f): 4171 ms
 -> upload Exercises (#0a0cbcd6-1abb-49f0-898a-595f606fc0bd): 4240 ms
 -> upload Exercises (#-7): 4482 ms
 -> upload Exercises (#3d1043a8-a2b1-434c-9b9d-c828f1f68f0a): 4635 ms
 -> upload Exercises (#05240b14-54ba-4dfb-8fb0-b42e043d9dce): 4709 ms
 -> upload Exercises (#cfe0d63b-3c8b-4be6-92b6-100e743665e1): 4844 ms
 -> upload Exercises (#7ad4b975-046f-4406-8b04-25ac28da811b): 4951 ms
 -> upload Exercises (#5a6126b2-a739-451c-9970-3bf699118b44): 5135 ms
 -> upload Exercises (#5ba2e6c7-4251-4a69-8653-36803b3bb691): 5188 ms
 -> upload Exercises (#1be8fa42-cc92-4b83-86c7-9ac827655a80): 5343 ms
 -> upload Exercises (#b091f92b-c71b-4fd8-989a-44c3866cd991): 5385 ms
 -> upload Exercises (#924c0b6d-8feb-4851-9f12-1fc385ff7b92): 5536 ms
 -> upload Exercises (#-12): 5692 ms
 -> upload Exercises (#d581583c-fccd-4e0e-b1bc-8b1e283d996b): 5781 ms
 -> upload Exercises (#7c9ac102-f90c-4550-9d72-48811a09f879): 5863 ms
 -> upload Exercises (#7efc18ec-c006-439f-9627-e79973e971b1): 6044 ms
 -> upload Exercises (#0bcd36ae-13cd-4c20-9083-66ab0f03b64d): 6351 ms
 -> upload Exercises (#70dc619d-4453-4bd5-89de-e6072636fe5a): 6912 ms
 -> upload Exercises (#5c544356-d2a9-472d-980e-36e0c9ba6361): 6930 ms
 -> upload Exercises (#bd113ba8-7320-45e5-ae4c-9c2c65f41bb3): 6973 ms
 -> upload Exercises (#ed3c6497-9bef-4c74-89eb-e999231189a0): 7061 ms
 -> upload Exercises (#ca9f6b49-80c5-47b7-923e-afe9bc81243b): 7097 ms
 -> upload Exercises (#bbadde78-2d8d-409a-8995-a241d61ffa98): 7263 ms
 -> upload Exercises (#ef813edc-3af5-4716-a58c-14690efdd231): 7418 ms
 -> upload Exercises (#443509fe-4274-413e-9339-65572b2dc519): 7511 ms
 -> upload Exercises (#bb9333c4-a432-4f62-b0ce-a1ca0fed749d): 7699 ms
 -> upload Exercises (#a04abb70-d6d3-44ae-9a7b-95fdf5e1276b): 7871 ms
 -> upload Exercises (#c6942e25-6cdf-4d90-b4b6-f2a78957f68e): 8022 ms
 -> upload Exercises (#7f967add-f683-423f-9315-1f9002868369): 8314 ms
 -> upload Exercises (#37426405-728f-4dd4-8b56-1e544efb0ae0): 8403 ms
 -> upload Exercises (#781e242e-dd21-4ff0-be2d-8380fe2ba7c2): 8403 ms
 -> upload Exercises (#12e538d3-a810-4782-86c5-94577dbaca9e): 8640 ms
 -> upload Exercises (#b79ab123-64b7-4406-8565-63bb6665ba4e): 8858 ms
 -> upload Exercises (#2578b5a3-cc34-4aa6-ad94-a9200cb4e2ed): 9076 ms
 -> upload Exercises (#8cb7b770-057e-4050-b3db-2b86d216fec0): 9213 ms
 -> upload Exercises (#b375e4c3-8b14-4d56-9edd-75d9d65c13aa): 9464 ms
 -> upload Exercises (#8e2a7fa9-3be6-4d8f-bf46-3d92071c2ea6): 9755 ms
 -> upload Exercises (#bdd9416b-3cda-4e3d-9b43-06ae47ca0a56): 9809 ms
 -> upload Exercises (#01c172dd-ab7e-4034-9445-70702a97ceac): 9879 ms
 -> upload Exercises (#bc0057ee-952b-4e78-a693-de911baff72b): 9900 ms
 -> upload Exercises (#2624665c-0598-4dfa-a23b-855b8c8d4665): 10207 ms
 -> upload Exercises (#54c3816d-feb0-4ee2-bc28-dc2173e94a0a): 10293 ms
 -> upload Exercises (#6bb31df9-397c-467d-9c7e-d4365e9840e9): 10350 ms
 -> upload Exercises (#34fa087c-7324-4382-a9d8-b181ba20e328): 10488 ms
 -> upload Exercises (#27088840-599a-47f0-9cc2-2e1a1d928016): 10693 ms
 -> upload Exercises (#23801d55-e1ce-48ff-a9b4-a79d166b77c3): 10762 ms
 -> upload Exercises (#6cdc9f1e-c247-4a30-b6fb-8d34e1a3eecf): 10868 ms
 -> upload Exercises (#425c5551-b090-44af-a317-e8f544e44572): 10991 ms
 -> upload Exercises (#a662b491-4f6d-41e0-b967-3a3d88e96972): 11081 ms
 -> upload Exercises (#02fe962f-2a13-4ae1-a1d5-86fb9566029f): 11218 ms
 -> upload Exercises (#4abf91ca-7c24-49c4-acd8-b54bc681bb78): 11424 ms
 -> upload Exercises (#5f27134d-d8dc-4f74-bfd8-58aa24964484): 11535 ms
 -> upload Exercises (#0905afe3-82a4-4c09-a2f1-04e30d4807b8): 11626 ms
 -> upload Exercises (#c5b1200d-2fd7-4377-a294-0484fc8826a5): 11680 ms
 -> upload Exercises (#a2720766-8b69-42bf-93c2-5f413ac02fdf): 11728 ms
 -> upload Exercises (#f20197ab-dc07-436f-b154-93c8597096d7): 11954 ms
 -> upload Exercises (#bd8106f2-46b7-4bfa-affd-a1484abbbd4c): 12160 ms
 -> upload Exercises (#f3d82074-1827-4610-8434-5efac78ca886): 13091 ms
 -> upload Exercises (#0e6667ac-7370-4a50-be46-5a90e321849d): 13182 ms
 -> upload Exercises (#5f60eb46-d0df-4eb2-97b8-c1ee56bccef4): 13303 ms
 -> upload Exercises (#fbaca3d6-75c3-4961-a9c3-76314329feb9): 13333 ms
 -> upload Exercises (#f6dd50e3-4cb9-4b7f-a469-fc59383ddad1): 13481 ms
 -> upload Exercises (#8333cf5c-e3b2-4a0a-b4d4-bbe235338c37): 13519 ms
 -> upload Exercises (#f0bd86ed-8a3f-416a-bf63-7820cf43614a): 13584 ms
 -> upload Exercises (#20284e72-4f1e-4f70-ba0d-345688332c2a): 13611 ms
 -> upload Exercises (#3f43cafe-809c-4996-ae29-2c2398928a7a): 13825 ms
 -> upload Exercises (#a4719012-0ce1-4268-8d8b-ccf423a04a58): 14123 ms
 -> upload Exercises (#72682efb-0c05-44b6-9f2e-32bd1a7679c0): 14463 ms
 -> upload Exercises (#a0235488-9a05-4889-99c1-5e553a4d2cf9): 14899 ms
 -> upload Exercises (#585260f2-ae5a-4ba4-a7c3-23292a9e189e): 14911 ms
 -> upload Exercises (#a5795802-ea5c-4180-8cd1-6a47b3b31ca7): 15380 ms
 -> upload Exercises (#f4fbf37b-f83e-4f22-865d-7b1be134a821): 15689 ms
 -> upload Exercises (#3f8f652d-72ac-4f35-9d57-79b336bbf7df): 16314 ms
 -> upload Exercises (#63ee445d-98f1-40ab-acc4-8d89f1ce6108): 17006 ms
 -> upload Exercises (#0ee5ae69-931f-407d-afb2-52767fa48d38): 17934 ms
 -> upload Exercises (#6247da70-9d31-4bc7-9f48-6d4a09f8f367): 19214 ms
 -> upload Exercises (#f1228ae5-6235-4cdd-92b5-1d3df5925105): 19664 ms
(Crash)
```

Just to give an idea of what kind of data is being uploaded. There are 1730 of these in my case, along with _some_ other (minor) data.
<img width="418" alt="image" src="https://user-images.githubusercontent.com/4264531/147633660-33390e55-e390-4894-a9e3-4d029a9a9143.png">

**Comment by argzdev** [comment]
Hi @zoltish, thanks for reporting. We'll try to investigate this and see what we can find. In the meantime, if you could provide us a [MCVE](https://stackoverflow.com/help/minimal-reproducible-example), it'll greatly help us investigate this further, thanks!

**Comment by argzdev** [comment]
Hi @zoltish, I created a clean project and tried inserting around ~2000 items with similar data structure to the one you provided with `setPersistenceEnabled(true)`, however I'm unable to experience any crash. 

Steps:
1. Sign in anonymously
2. Turn on airplane mode
3. Insert around ~2000 data
3. Turn off airplane mode
4. App begins to sync, however app does not crash

I tried 2 ways of inserting data, relevant code:
```
fun add(){
     for(i in 0..2000){
         //writeWorkOut1()
         writeWorkOut2()
     }
}

fun writeWorkOut1(){
    database.child("workout").child("testId").push().setValue(createWorkOut())
}
    
fun writeWorkOut2(){
    val key = database.child("workout").child("testId").push().key
    if (key == null) {
        Log.w(TAG, "Couldn't get push key for posts")
        return
    }

    val workout = createWorkOut()
    val workoutValues = workout.toMap()

    val childUpdates = hashMapOf<String, Any>(
        "/workout/$key" to workoutValues
    )

    database.updateChildren(childUpdates)
}
```

With that said, can you help me with the following implementations?
- Implementation of measuring time of each process (I haven't added measuring time for each processes)
- Implementation in uploading multiple data

Thanks!

**Comment by zoltish** [comment]
Thanks @argzdev! You mention turning airplane mode on, is there a particular reason for that?

My implementation when uploading multiple data is similar to your `writeWorkOut1` (and performing it for each value). I am however using the playservices extension to coroutines, for the .await functionality to tasks: `reference.setValue(data).await()`

I just added the timing measurement to hopefully narrow down where the issue is; Ill link the function below. In my example above I just wrapped the `reference.setValue()` calls.

I hope this helps? Please let me know if youd like me to share more code! As an aside, when performing the download & upload task, I do use `coroutineScope{}` and map each call into a Deferred (by using `async {}`) and then `.awaitAll()`. I dont know if that makes a difference, but do let me know if that raises any suspicion!

```
inline fun <R> logTiming(
    tag: String,
    min: Long = 0,
    block: () -> R,
): R {
    val start = currentTimeMillis()
    val result = block()
    val end = currentTimeMillis()
    val duration = end - start
    if (duration >= min) {
        logcat { "$tag: $duration ms" }
    }
    return result
}
```

**Comment by argzdev** [comment]
Hi @zoltish, thanks for the explanation. I turned airplane mode on/off as an experiment to see if issue might be due to the surge of local data being uploaded during sync, however, this doesn't seem to be the case.

As for the implementation you've provided. Here's what I tried:
```
GlobalScope.launch {
    val deferredList = (1..4000).map {
        async {
            val start = currentTimeMillis()
            sharedViewModel.writeWorkOut2(database).await()
            val end = currentTimeMillis()
            Log.d(TAG, "add $it: ${end - start} ms")
        }
    }
    deferredList.awaitAll()
    Log.d(TAG, "add: done")
}

fun writeWorkOut2(database: DatabaseReference) = GlobalScope.async {
    database.child("workout").child("testId").push().setValue(createWorkOut())
}
```

I tried uploading around ~4000 data, but I still don't experience any crashing.
```
2022-01-01 02:30:07.413 8279-8343/com.argz.issue3276 D/MainFragment: add 3995: 8 ms
2022-01-01 02:30:07.413 8279-8344/com.argz.issue3276 D/MainFragment: add 3993: 16 ms
2022-01-01 02:30:07.417 8279-8341/com.argz.issue3276 D/MainFragment: add 3994: 13 ms
2022-01-01 02:30:07.417 8279-8342/com.argz.issue3276 D/MainFragment: add 3991: 23 ms
2022-01-01 02:30:07.417 8279-8342/com.argz.issue3276 D/MainFragment: add 3987: 37 ms
2022-01-01 02:30:07.417 8279-8344/com.argz.issue3276 D/MainFragment: add 3999: 4 ms
2022-01-01 02:30:07.419 8279-8342/com.argz.issue3276 D/MainFragment: add 3996: 6 ms
2022-01-01 02:30:07.422 8279-8343/com.argz.issue3276 D/MainFragment: add 3997: 9 ms
2022-01-01 02:30:07.429 8279-8341/com.argz.issue3276 D/MainFragment: add 3998: 16 ms
2022-01-01 02:30:07.445 8279-8342/com.argz.issue3276 D/MainFragment: add: done
```

Perhaps the issue is somewhere else? Or am i missing anything?

**Comment by zoltish** [comment]
I really appreciate the prompt responses @argzdev! I believe Ive narrowed down the issue even further, and if my understanding is correct the issue is probably not related to firebase directly. 

My application also synchronizes with Google Fit, and during this rather heavy import - all of these sessions are also being inserted into Google Fit, pretty much simultaneously as theyre being uploaded to the RTDB. 

I cant really say why that would lead to issues like this, considering that both are performed in the background, etc, although I am using the same kind of `map async followed by awaitAll` flow. Ill rethink and rework my implementations over the coming days to see if that makes any difference and report back once I know! 

As for this issue, what youre doing looks pretty much identical to what Im doing. The only key difference left (beyond the Google Fit thing) is that Im also performing these tasks in parallell for exercises, training-programs, etc. Im wondering if the simultaneous uploads to different databaseReferences could cause any of the issues? All of this seems to work perfectly fine for smaller datasets, even with the Google Fit integration doing its thing. My entire dataset thats being uploaded is comprised of about 20+2000+2000+300+40+1 data structures, with approximately the same data complexity as the one I linked above!

**Comment by argzdev** [comment]
Thanks for the detailed explanation, @zoltish. Your app does seem to do a lot of complex processes simultaneously. Perhaps you could try isolating each processes and do tests (e.g. like what you did with logging) on those processes to see which is under performing, and if these tests perform properly. It may be plausible that the simultaneous uploads are causing issues. For what's it worth, using our [profiler tool](https://firebase.google.com/docs/database/usage/monitor-performance#use_the_profiler_tool) may help you identify the underlying issue. For now, let’s keep this issue open until you have further details or information. Thanks!

**Comment by zoltish** [comment]
I think the actual problem is that if I fire a lot of requests simultaneously, a lot of them will come through quickly - but others will lag behind, and after a while they will accumulate and slow down everything enough such that new requests are just added to the pile and almost never finishing (initially ~350 ms, later ~30s) - until the OOM error happens. 

If that is correct, you should see something similar in your example if you perform all those requests multiple times in parallell, which is somewhat identical to what Im doing:

Plates -> 50
Profiles -> 1
Exercises -> 420
Workouts -> 3000
Sessions -> 3000
Programs -> 250

All of these are being uploaded simultaneously, and within each task every item is also being uploaded simultaneously. 

The workaround Ive experimented with is using `list.chunked(150)` and then using async on the 150 items, instead of doing it on each individual item. I guess the goal now is to find the appropiate number to chunk them in for best efficiency, 150 works great though, and I hope this illustrates the issue well! I dont know if its strictly related to Firebase or not, this could probably happen in other scenarios as well - but this is the first scenario where Ive come across it!

**Comment by argzdev** [comment]
Hi @zoltish, I tried firing multiple requests using a deferredList, but the uploads do not seem to be affected with simultaneous requests, I got the same result similar to my previous code snippet. To try this, what I did was create a dynamic function of inserting data in the db. I used the `createWorkOut` function as a data input for all the inserts for simplicity.

Relevant code:
```
fun writeOnDb(database: DatabaseReference, path: String) = GlobalScope.async {
   database.child(path).child("testId").push().setValue(createWorkOut())
}

fun cleanDb(database: DatabaseReference, path: String) = GlobalScope.async {
   database.child(path).setValue(null)
}
```

I’ve added a `cleanDb` to clear out the db every time I add data. As for the parallel requests:
```
fun add(){
    GlobalScope.launch {
            sharedViewModel.cleanDb(database, "plate").await()
            sharedViewModel.cleanDb(database, "profile").await()
            sharedViewModel.cleanDb(database, "exercise").await()
            sharedViewModel.cleanDb(database, "workout").await()
            sharedViewModel.cleanDb(database, "session").await()
            sharedViewModel.cleanDb(database, "program").await()

            val deferred1 = async {
                (1..50).map {
                    val start = currentTimeMillis()
                    sharedViewModel.writeOnDb(database, "plate").await()
                    val end = currentTimeMillis()
                    Log.d(TAG, "plate $it: ${end - start} ms")
                }
            }
            val deferred2 = async {
                val start = currentTimeMillis()
                sharedViewModel.writeOnDb(database, "profile").await()
                val end = currentTimeMillis()
                Log.d(TAG, "profile: ${end - start} ms")
            }
            val deferred3 = async {
                (1..420).map {
                    val start = currentTimeMillis()
                    sharedViewModel.writeOnDb(database, "exercise").await()
                    val end = currentTimeMillis()
                    Log.d(TAG, "exercise $it: ${end - start} ms")
                }
            }
            val deferred4 = async {
                (1..3000).map {
                    val start = currentTimeMillis()
                    sharedViewModel.writeOnDb(database, "workout").await()
                    val end = currentTimeMillis()
                    Log.d(TAG, "workout $it: ${end - start} ms")
                }
            }
            val deferred5 = async {
                (1..3000).map {
                    val start = currentTimeMillis()
                    sharedViewModel.writeOnDb(database, "session").await()
                    val end = currentTimeMillis()
                    Log.d(TAG, "session $it: ${end - start} ms")
                }
            }
            val deferred6 = async {
                (1..250).map {
                    val start = currentTimeMillis()
                    sharedViewModel.writeOnDb(database, "program").await()
                    val end = currentTimeMillis()
                    Log.d(TAG, "program $it: ${end - start} ms")
                }
            }
            val deferredList = listOf(deferred1, deferred2, deferred3, deferred4, deferred5, deferred6)
            deferredList.awaitAll()
            Log.d(TAG, "done")
        }
}
```

Based on our RTDB [limits](https://firebase.google.com/docs/database/usage/limits#global), there shouldn’t be any problem with simultaneous requests, unless the limits are being hit. May I ask if your write/update events are greater than 1 MB? Or is it possible that the way you're processing data before uploading it is causing issues?

**Comment by zoltish** [comment]
Hello @argzdev! 

Im using `Dispatchers.IO`; not sure if it actually matters - but the GlobalScope uses `Dispatchers.Default` so I thought Id mention it.

What kind of timings are you seeing in the `end - start` calculations in your code? Is it normal to see ~350 ms per call to `.await()`? That number usually grows to ~1500 ms when performing a lot of work as well for me, just curious whether or not Im doing something wrong there. It takes roughly 60-120 seconds to perform the entire upload of one "branch" with 3000 items in my case.

My brain is running on fumes at this point, but I think the end result of what Im doing would be identical to your code snippet using something along the lines of (notice the extra `async await`):

```
   val deferred6 = async {
                (1..250).map { async {
                    val start = currentTimeMillis()
                    sharedViewModel.writeOnDb(database, "program").await()
                    val end = currentTimeMillis()
                    Log.d(TAG, "program $it: ${end - start} ms")
                }
                }.awaitAll()
            }
```

So each branch runs its upload in parallel, and each item upload in parallel as well. 

I dont think my updates are >1MB; how can I tell? In my example project Im using 10 MB of storage in total, but considering the amount of items - I think each update is comprised of <1MB.

**Comment by argzdev** [comment]
Hi @zoltish, thanks for the extra details. ~350 ms is, I think, within the acceptable range, however this can vary depending on your internet connection.

For the speed calculations, I see around 10+ ms to at most 20+ ms only.

Here is the beginning:
```
2022-01-07 22:15:58.638 30511-30597/com.argz.issue3276 D/MainFragment: plate 1: 20 ms
2022-01-07 22:15:58.638 30511-30596/com.argz.issue3276 D/MainFragment: program 1: 8 ms
2022-01-07 22:15:58.644 30511-30596/com.argz.issue3276 D/MainFragment: program 2: 6 ms
2022-01-07 22:15:58.644 30511-30597/com.argz.issue3276 D/MainFragment: plate 2: 6 ms
2022-01-07 22:15:58.649 30511-30599/com.argz.issue3276 D/MainFragment: profile: 21 ms
2022-01-07 22:15:58.651 30511-30597/com.argz.issue3276 D/MainFragment: plate 3: 7 ms
2022-01-07 22:15:58.652 30511-30598/com.argz.issue3276 D/MainFragment: exercise 1: 17 ms
2022-01-07 22:15:58.652 30511-30596/com.argz.issue3276 D/MainFragment: program 3: 8 ms
```
And then around the end:
```
2022-01-07 22:16:14.656 30511-30597/com.argz.issue3276 D/MainFragment: workout 2968: 17 ms
2022-01-07 22:16:14.656 30511-30596/com.argz.issue3276 D/MainFragment: session 3000: 1 ms
2022-01-07 22:16:14.660 30511-30597/com.argz.issue3276 D/MainFragment: workout 2969: 4 ms
2022-01-07 22:16:14.673 30511-30597/com.argz.issue3276 D/MainFragment: workout 2970: 12 ms
2022-01-07 22:16:14.675 30511-30597/com.argz.issue3276 D/MainFragment: workout 2971: 2 ms
2022-01-07 22:16:14.679 30511-30599/com.argz.issue3276 D/MainFragment: workout 2972: 4 ms
2022-01-07 22:16:14.684 30511-30599/com.argz.issue3276 D/MainFragment: workout 2973: 5 ms
```

My guess is that if each of your updates are less than 1 MB, then most likely the exceeding the write limit isn't the issue. Perhaps as a test, you could try isolating each processes(e.g. workout, session, etc.) and do tests separately and not simultaneously, just to check if those processes are performing much better if they're run by themselves.

If there's any chance you could provide us a minimal reproducible example, this can help us speed up the investigation.

**Comment by zoltish** [comment]
Hey @argzdev! 

Got it. Im located in Sweden with a connection of ~250 mbps at the time of testing, using _Database location: United States (us-central1)_.

**I no longer know if this is a real issue considering that everything works just fine if there arent a ton of tasks being ran in parallel.** 

I did however reduce my implementation even further, and while its not a MCVE, I hope it can paint the full picture of what eventually leads to the crash above. Ironically, running this code alone doesnt crash, I suspect Id need about 2000 items for that (usually it crashes around 32 000 ms); but I can see why it would crash in a scenario where other tasks are ran in parallel to this. 

```
    GlobalScope.launch {
            val adapter = registry[Session::class]!!

            //Basically database.getReference("sessions").child("userId")
            val path = token.getPath(adapter.branch)

            adapter
                .all() // 1730 Sessions
                .map { model ->
                    async {
                        val data = adapter.wrap(model)

                        //1500 ms -> 24000 ms
                        logTiming("upd ${model.id()}") {
                            //reference.child("id").setValue(data).await()
                            path
                                .child(adapter.key(model))
                                .update(data)
                        }
                    }
                }
                .awaitAll()

            logcat { "done" }
        }
```

Uploading all the data sequentially literary takes forever, which is why I started experimenting with running tasks in parallel. With this crash in mind, and the performance regressions it can introduce, I dont really know where the sweet spot is. Im testing this on a Pixel 3, and lower end devices might suffer the regressions earlier on. The safest approach I can think of currently is to chunk 25-50 items into an async call, speeding up large uploads (in practice, likely only happening on the initial sign-in) while allowing smaller uploads to be ran in parallel as well. 

I dont know if theres any more info that I can/should provide? At the end of the day this feels like a non issue now, but please let me know if you disagree. I would love to provide a MCVE but I just dont have the time to do so right now as Im still in the middle of a big release. Feel free to close this issue or ask follow-up questions if they can be of help!

**Comment by argzdev** [comment]
Hi @zoltish, thanks for the update and providing extra details. The approach you're using does seem like the best option you can do for now. With that said, I'll be closing this issue. If you ever find any new details that points to our RTDB causing this issue, feel free to comment back here or open a new issue. Thanks again for the your report, this has been an interesting insight!
